bolt/src/Passes/Instrumentation.cpp

```diff
@@ -177,15 +177,43 @@ Instrumentation::createInstrumentationSnippet(BinaryContext &BC, bool IsLeaf) {
   Label = BC.Ctx->createNamedTempSymbol("InstrEntry");
   Summary->Counters.emplace_back(Label);
   std::vector<MCInst> CounterInstrs;
-  CounterInstrs.resize(IsLeaf? 5 : 3);
   uint32_t I = 0;
+
+  if (BC.isX86() && !opts::ConservativeInstrumentation) {
+    CounterInstrs.resize(IsLeaf ? 13 : 11);
+  } else {
+    CounterInstrs.resize(IsLeaf? 5 : 3);
+  }
+
   // Don't clobber application red zone (ABI dependent)
   if (IsLeaf)
     BC.MIB->createStackPointerIncrement(CounterInstrs[I++], 128,
                                         /*NoFlagsClobber=*/true);
-  BC.MIB->createPushFlags(CounterInstrs[I++], 2);
-  BC.MIB->createIncMemory(CounterInstrs[I++], Label, &*BC.Ctx);
-  BC.MIB->createPopFlags(CounterInstrs[I++], 2);
+  if (BC.isX86() && !opts::ConservativeInstrumentation) {
+    // performance improvement based on the LLVM optimziation
+    // https://reviews.llvm.org/D6629
+    // LAHF/SAHF are used instead of PUSHF/POPF
+    const MCPhysReg RAX = BC.MIB->getRAXRegister(8);
+    const MCPhysReg AL = BC.MIB->getRAXRegister(1);
+    // PUSHF
+    BC.MIB->createPushRegister(CounterInstrs[I++], RAX, 8);
+    BC.MIB->createClearRegisterByMovInstr(CounterInstrs[I++], RAX, 8);
+    BC.MIB->createLahf(CounterInstrs[I++]);
+    BC.MIB->createPushRegister(CounterInstrs[I++], RAX, 8);
+    BC.MIB->createClearRegisterByMovInstr(CounterInstrs[I++], RAX, 8);
+    BC.MIB->createSaveOVFlagToRegister(CounterInstrs[I++], AL);
```

Comment:
In general, the MCPlusBuilder interface should be target-agnostic and reflect that the user of these functions may be writing code to either X86 or Arm. But this is not always possible, so functions that are exceptions to this rule and are X86-only should be explicitly marked as "X86". So, in this case, something like "createX86SaveOVFlagToRegister" should be better.

But because this whole code sequence is pretty specific to X86 too, I think an idea that will work better in terms of design is to put the entire code sequence behind X86MCPlusBuilder through a function called "createX86FlagsPreservingIncMemory()", which returns the full sequence from Instrumentation.cpp:196 to :211, instead of adding these smaller building blocks to the MCPlusBuilder interface.  The idea of this interface is that target-specific code generation should be done in target-specific classes, instead of being done directly in the target-agnostic classes such as Instrumentation.cpp. If we move this logic to X86MCPlusBuilder, I think the design will be cleaner.